### PR TITLE
Include ANIm self vs self in test fixtures; log ANIm from snakemake

### PR DESCRIPTION
**Pull request by peterjc**
Continues work for https://github.com/pyani-plus/pyani-plus/issues/4 to run nucmer for the ANIm method. Builds on #128, and concludes with logging each a single ANIm comparison from snakemake following the pattern used for ANIb and fastANI.

We need to expand the tests to include self-vs-self in the ANIm targets. This is mostly straightforward, other than an oddity with pyANI v0.3 putting 1 on the diagonal for sim_errors.

Required a tweak for getting the query coverage to match pyANI v0.3, which was down to the subject vs reference ordering discussed on #128 and much earlier at https://github.com/widdowquinn/pyani/issues/151

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] This change requires a documentation update
- [ ] This is a documentation update

## Action Checklist

- [x] Work on a single issue/concept (if there are multiple separate issues to address, please use a separate pull request for each)
- [ ] Fork the `pyani-plus` repository under your own account (please [allow write access for repository maintainers](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))
- [x] Set up an appropriate development environment (please see `CONTRIBUTING.md` for this repository)
- [x] Create a new branch with a short, descriptive name
- [x] Work on this branch
  - [x] style guidelines have been followed (please see `CONTRIBUTING.md` for this repository)
  - [x] new code is commented, especially in hard-to-understand areas
  - [x] corresponding changes to documentation have been made
  - [x] tests for the change have been added that demonstrate the fix or feature works
- [x] Test locally with `pytest-plus -v` **non-passing code will not be merged**
- [x] Rebase against `origin/master`
- [x] Check changes with linting/formatting tools before submission (please see `CONTRIBUTING.md` for this repository)
- [x] Commit branch
- [x] Submit pull request, describing the content and intent of the pull request
- [x] Request a code review
- [x] Continue the discussion at [`Pull requests` section](https://github.com/widdowquinn/pyani-plus/pulls) in the `pyan-plus` repository

### CODE REVIEW

**Comment by peterjc** [comment]
The ones on the ANIm sim-errors diagonal from pyANI v0.3 was the last hurdle in getting the matrices recorded via snakemake to match the expected output - final part of this pending (could add it to this branch, or make another followup PR shortly).

**Comment by peterjc** [comment]
Note we might alternatively fix pyANI v0.3 to put zeros on the sim-errors diagonal...

**Comment by peterjc** [comment]
It is a one character fix on the pyani side, so I lean to fixing the output there - rather the somewhat fiddly fix proposed here post-processing the output in the fixture generator script.

**Comment by peterjc** [comment]
My plan is once https://github.com/widdowquinn/pyani/pull/439 is merged, I will drop the diagonal fixes from this branch, and re-generate the matrices using the updated pyani instead.

File: pyani_plus/workflows/snakemake_anim.smk [hunk 1]
```diff
@@ -56,11 +56,21 @@ rule delta:
...
         delta_filter=config["delta_filter"],
         outdir=config["outdir"],
     input:
         "{outdir}/{genomeA}_vs_{genomeB}.delta",
+        genomeA=get_genomeA,
+        genomeB=get_genomeB,
     output:
         "{outdir}/{genomeA}_vs_{genomeB}.filter",
     shell:
-        "{params.delta_filter} -1 {input} > {wildcards.outdir}/{wildcards.genomeA}_vs_{wildcards.genomeB}.filter"
```
Comment (peterjc) [inline pyani_plus/workflows/snakemake_anim.smk:66]: Can't use ``{input}`` here anymore as that now expands to three filenames (the delta file and two FASTA files).

Comment (kiepczi) [reply pyani_plus/workflows/snakemake_anim.smk:66]: I'm happy with the change if it works and make things faster/better. I'm just curious what change triggered this... just so I know for the future :)

Here `{input}` refered to `snakemake` input varaiable, in this case `{outdir}/{genomeA}_vs_{genomeB}.delta`,  so even if other files are present there, it shouldn't matter. Or I misunderstood something here.

File: pyani_plus/private_cli.py [hunk 2]
```diff
@@ -440,8 +440,9 @@ def log_anim(
     # after the computation has finished (on the same machine)
     nucmer_tool = tools.get_nucmer()
 
-    r_aligned_bases, q_aligned_bases, identity, sim_errors = method_anim.parse_delta(
-        deltafilter
+    # Note inconsistency in how nucmer uses query vs reference and flips order
+    query_aligned_bases, subject_aligned_bases, identity, sim_errors = (
+        method_anim.parse_delta(deltafilter)
     )
```
Comment (kiepczi) [inline pyani_plus/private_cli.py:446]: I don't think this change is needed. `method_anim.parse_delta` reports values in the following order:

```python
    return (
        get_aligned_bases_count(regions_ref),
        get_aligned_bases_count(regions_qry),
        avrg_identity,
        sim_error,
    )
```

Comment (kiepczi) [reply pyani_plus/private_cli.py:446]: What I think needs to be changed is line 449, as I think this is flipped
```
used_query, used_subject = deltafilter.stem.split("_vs_")
```

Comment (kiepczi) [reply pyani_plus/private_cli.py:446]: For context... `rule_delta` runs the following command:
```python
{params.nucmer} -p {wildcards.outdir}/{wildcards.genomeA}_vs_{wildcards.genomeB} --mum {input.genomeA} {input.genomeB}
```
According to nucmer documentation, they refer to them as reference (genomeA) and query (genomeB)

```bash
USAGE: nucmer  [options]  <Reference>  <Query>
```

`rule_delta` uses the same format eg. genomeA_vs_genomeB:

```python
{outdir}/{genomeA}_vs_{genomeB}.filter
```

**Comment by kiepczi** [comment]
Hi @peterjc. My apologies for missing this code review ealier.

I think there is only one thing that needs to be adressed before this branch can be merged. Hope my comments make sense :)

File: pyani_plus/workflows/snakemake_anim.smk (hunk 1 above)
Comment (peterjc) [reply pyani_plus/workflows/snakemake_anim.smk:66]: Originally there was only one input (the delta file).

Now there are three input files, the delta file (given to delta filter as before), plus the query and subject files (needed for the DB logging).

This is why ``{input}`` no longer gives just the delta filename, but three file names.

File: pyani_plus/private_cli.py (hunk 2 above)
Comment (peterjc) [reply pyani_plus/private_cli.py:446]: Ah! I understood we had the convention in pyANI plus (and legacy pyANI), that the intermediate files were all called ``<query>_vs_<subject>.<extension>`` and strongly feel that should be used consistently.

It seems then that we should swap this:

``{params.nucmer} -p {wildcards.outdir}/{wildcards.genomeA}_vs_{wildcards.genomeB} --mum {input.genomeA} {input.genomeB}``

and:

```python
query_aligned_bases, subject_aligned_bases, identity, sim_errors = (
    method_anim.parse_delta(deltafilter)
)
```

to be this:

``{params.nucmer} -p {wildcards.outdir}/{wildcards.genomeA}_vs_{wildcards.genomeB} --mum {input.genomeB} {input.genomeA}``

and:

```python
subject_aligned_bases, query_aligned_bases, identity, sim_errors = (
    method_anim.parse_delta(deltafilter)
)
```

Then it should all be consistent?

Comment (kiepczi) [reply pyani_plus/private_cli.py:446]: If we go this rount we will need to change this for `dnadiff` method too.

**Comment by peterjc** [comment]
Do you want to postpone query-vs-subject order to a follow up PR, especially given this likely affects dnadiff as well?

**Comment by kiepczi** [comment]
Whichever works best. If you're busy with other tasks, we can merge this as-is, and I'll fix the `<query>_vs_<subject>` in the next PR for both methods. That way, I can move on to working on #131.

**Comment by peterjc** [comment]
I'm taking your last comment as an implicit PR approval, merging now.

**Pull request merged**